src/ui/uiMain.ts

```diff
@@ -44,3 +44,6 @@ export class UiMain {
     return this.ipcUiService.workerInvoke(message);
   }
 }
+
+export const uiMainInstance = new UiMain(new IpcUiService());
+uiMainInstance.main();
```

Comment:
Should we instantiate this instead in `GlobalStateProvider`?

Reply:
It depends. Can we guarantee that it will absolutely never be instantiated twice?

Reply:
The component could re-render, but it shouldn't happen often as it is very high up in the chain of components and only 1 instance would be used by the app at a given time. 
- Is it important that the exact same instance lives the whole time the ui is open?
- If not, in the case of a re-render, would we have to perform some cleanup to make sure that any previous instance are closed (if there's any kind of active connection)?

I think [`useRef`](https://reactjs.org/docs/hooks-reference.html#useref) could also be another way to achieve this:
> The returned object will persist for the full lifetime of the component.

> useRef will give you the same ref object on every render.

Reply:
https://dev.to/salehmubashar/useref-or-usestate-which-is-better-258j

> References do not affect component rendering but states do.

> When the reference value is changed, it is updated without the need to refresh or re-render. However in useState, the component must render again to update the state or its value.

Reply:
ok let's try that

Reply:
> Is it important that the exact same instance lives the whole time the ui is open?

Not terrible, but there is useless init logic that will be re-run if we re-initiate it.

Reply:
Meh. Taking a look at it again I don't really know what to do with the default value.

Reply:
I'll merge like this for now, but if you have a solution that works well by creating it from the component then feel free to open a PR.

IMO the default value from providers should be a callback. It would be much simpler.